spec/Candid.md

```diff
@@ -1318,9 +1314,12 @@ Deserialisation at an expected type sequence `(<t'>,*)` proceeds by
 
  * checking for the magic number `DIDL`
  * using the inverse of the `T` function to decode the type definitions `(<t>,*)`
- * check that `(<t>,*) <: (<t'>,*)`, else fail
  * using the inverse of the `M` function, indexed by `(<t>,*)`, to decode the values `(<v>,*)`
- * use the coercion function `C[(<t>,*) <: (<t'>,*)]((<v>,*))` to understand the decoded values at the expected type.
+ * use the coercion function `v : t ~> v' : t'` to understand the decoded values at the expected type.
+
+Note on implementation:
```

Comment:
So I guess this is admitting defeat for having an abstract cost model?
I can see that the formulation using inference rules makes this hard.
Would it still be hard if we specified the decoder as a monadic function, returning a cumulative cost and optional value, making the backtracking and skips explicit?

Reply:
Or would that be too ugly or prescriptive?

Reply:
I think it's possible to have an abstract cost model by assigning weights for each derivation rule. But I see several problems with this approach: 
1) It forces the implementation to strictly follow the derivation rules; 
2) Types in the host language is not completely captured by the model. For example, `[Nat8]` vs `Blob` in Motoko, and `Vec<(K, V)>` vs `HashMap<K,V>` in Rust. They map to the same Candid type, but have different cost in the host language. Even with the same `record` type, it can be more expensive to construct a record in Rust than in Motoko (or vice versa);
3) Even if we have a cost model, it would be hard to come up with a threshold in the spec. It's application dependent (canister endpoints and inter-canister calls may need different bounds, and stable memory won't need any limit). Plus there is always a gap between the cost model and the real implementation.

Looking from a different angle, this is a resource problem originating from a specific platform and a specific implementation. Given infinite resource, the Candid spec doesn't need to reject these large payload. Just like we don't specify stack depth in the spec, we can leave the concrete cost metering to the implementation, so that it matches better with the platform and language runtime.